eras/byron/ledger/impl/src/Cardano/Chain/Genesis/Data.hs

```diff
@@ -93,7 +95,7 @@ instance MonadError SchemaError m => FromJSON m GenesisData where
     GenesisData
       <$> fromJSField obj "bootStakeholders"
       <*> fromJSField obj "heavyDelegation"
-      <*> fromJSField obj "startTime"
+      <*> (id <$!!> fromJSField obj "startTime")
```

Comment:
Despite that being what Alexey suggested in the ticket, I think this reads better:

```suggestion
      <*> (force <$> fromJSField obj "startTime")
```

I looked at the [source](https://hackage.haskell.org/package/deepseq-1.5.0.0/docs/src/Control.DeepSeq.html#%3C%24%21%21%3E) of `<$!!>` and I believe it's equivalent.

Reply:
Technically, the equivalence depends on the behaviour of the underlying monad, but the [monad laws](https://hackage.haskell.org/package/base-4.20.0.1/docs/Control-Monad.html#t:Monad) imply

```haskell
force <$> m = m >>= pure . force
force <$> m = m >>= \x -> pure (force x)
force <$> m = m >>= \x -> pure (x `deepseq` x)
force <$> m = m >>= \x -> pure (deepseq x x)
```

We also have

```haskell
id <$!!> m = m >>= \x -> pure $!! id x
id <$!!> m = m >>= \x -> (id x) `deepseq` pure (id x)
id <$!!> m = m >>= \x -> x `deepseq` pure x
id <$!!> m = m >>= \x -> deepseq x (pure x)
```

So it depends on whether `deepseq x (pure x)` is the same as `pure (deepseq x x)`. Technically it isn't, because the first one will evaluate `x` earlier, before the `pure` is executed, and the second one will evaluate `x` later, as the `pure` is executed. However, in an environment where the `pure` is going to be forced by monadic evaluation, it has the same effect, because we care only that the value is forced by the time the monadic evaluation is finished.

Reply:
OTOH, `<*>` forces the `error` (ie this program gets an exception before printing anything) so I think it's OK.

```haskell
import Control.Applicative
import Control.DeepSeq
import Data.Maybe

main :: IO ()
main = do
  let a = force <$> Just (error "a") :: Maybe Char
  let b = id <$!!> Just (error "b") :: Maybe Char

  print . isJust $ (,) <$> a <*> b
```

Reply:
@neilmayhew you did go above and beyond on this suggestion :smile:

Yeah, `force <$>` will do just fine, since it marks the whole field to be evaluated to NF and the field itself is strict, so as long as the `GenesisData` result of `fromJSON` is forced the `startTime` will be forced to NF.

@neilmayhew my version is one character shorter :stuck_out_tongue:, but I do agree that `force <$>` reads much better

Reply:
> @neilmayhew you did go above and beyond on this suggestion 😄

Yeah, I did! Since I've been working on the Alonzo trace tests problem I've realized that my understanding of laziness wasn't as deep as it should be, so I'm taking every opportunity I can to learn more, and I thought I might as well share what I'd learned about this one.